lib/concurrent/ivar.rb

```diff
@@ -114,14 +121,24 @@ def set(value)
     # @param [Object] reason for the failure
     # @raise [Concurrent::MultipleAssignmentError] if the `IVar` has already
     #   been set or otherwise completed
+    # @return [IVar] self
     def fail(reason = StandardError.new)
-      complete(false, nil, reason)
+      set { raise reason }
```

Comment:
Calling `#complete` here is incompatible with `Promise`. A promise can be set only once so calling `#fail` more than once needs to raise an exception. That doesn't happen when we call `#complete` here. It's an easy fix. I'll override `#fail` in `Promise` and return `IVar` to the original implementation.